File: src/test/java/com/fml/common/configuration/ConfigurationReaderTests.java
package com.fml.common.configuration;

import static org.assertj.core.api.Assertions.assertThat;

import java.math.BigDecimal;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.springframework.core.env.Environment;
import org.springframework.core.env.StandardEnvironment;
import org.springframework.core.env.MapPropertySource;

class ConfigurationReaderTests {

    private static Environment environment(String... keysAndValues) {
        Map<String, Object> values = new LinkedHashMap<>();
        for (int i = 0; i < keysAndValues.length; i += 2) {
            values.put(keysAndValues[i], keysAndValues[i + 1]);
        }
        StandardEnvironment environment = new StandardEnvironment();
        environment.getPropertySources().addFirst(new MapPropertySource("test", values));
        return environment;
    }

    @Test
    void valuesAreParsedWhenPresent() {
        Environment environment = environment(
                "Fml:Flag", "true",
                "Fml:Count", "42",
                "Fml:Ratio", "1.5",
                "Fml:Amount", "250000");

        assertThat(ConfigurationReader.getBool(environment, "Fml:Flag", false)).isTrue();
        assertThat(ConfigurationReader.getInt(environment, "Fml:Count", 0)).isEqualTo(42);
        assertThat(ConfigurationReader.getDouble(environment, "Fml:Ratio", 0)).isEqualTo(1.5);
        assertThat(ConfigurationReader.getDecimal(environment, "Fml:Amount", BigDecimal.ZERO))
                .isEqualByComparingTo(new BigDecimal("250000"));
    }

    @Test
    void fallbacksAreUsedForMissingAndUnparseableValues() {
        Environment environment = environment("Fml:Count", "not-a-number");

        assertThat(ConfigurationReader.getBool(environment, "Fml:Missing", true)).isTrue();
        assertThat(ConfigurationReader.getBool(environment, "Fml:Count", true)).isTrue();
        assertThat(ConfigurationReader.getInt(environment, "Fml:Count", 7)).isEqualTo(7);
        assertThat(ConfigurationReader.getDouble(environment, "Fml:Missing", 2.0)).isEqualTo(2.0);
        assertThat(ConfigurationReader.getDecimal(environment, "Fml:Missing", new BigDecimal("3")))
                .isEqualByComparingTo(new BigDecimal("3"));
    }

    @ParameterizedTest
    @ValueSource(strings = {"en-US", "de-DE", "fr-FR"})
    void numericValuesAreParsedIndependentlyOfTheCurrentLocale(String locale) {
        Locale previous = Locale.getDefault();
        Locale.setDefault(Locale.forLanguageTag(locale));
        try {
            Environment environment = environment(
                    "Fml:Ratio", "1.5",
                    "Fml:Amount", "250000.75");

            assertThat(ConfigurationReader.getDouble(environment, "Fml:Ratio", 0)).isEqualTo(1.5);
            assertThat(ConfigurationReader.getDecimal(environment, "Fml:Amount", BigDecimal.ZERO))
                    .isEqualByComparingTo(new BigDecimal("250000.75"));
        } finally {
            Locale.setDefault(previous);
        }
    }

    @Test
    void groupedDecimalsAreParsedLikeTheInvariantCulture() {
        Environment environment = environment("Fml:Amount", "250,000.75");

        assertThat(ConfigurationReader.getDecimal(environment, "Fml:Amount", BigDecimal.ZERO))
                .isEqualByComparingTo(new BigDecimal("250000.75"));
    }
}

File: tests/Fml.Common.Tests/ConfigurationReaderTests.cs
using System.Globalization;
using FML.Common.Configuration;
using Microsoft.Extensions.Configuration;

namespace Fml.Common.Tests;

public class ConfigurationReaderTests
{
    private static IConfiguration Configuration(params (string Key, string Value)[] values) =>
        new ConfigurationBuilder()
            .AddInMemoryCollection(values.Select(v => new KeyValuePair<string, string?>(v.Key, v.Value)))
            .Build();

    [Fact]
    public void Values_are_parsed_when_present()
    {
        var configuration = Configuration(
            ("Fml:Flag", "true"),
            ("Fml:Count", "42"),
            ("Fml:Ratio", "1.5"),
            ("Fml:Amount", "250000"));

        Assert.True(configuration.GetBool("Fml:Flag", false));
        Assert.Equal(42, configuration.GetInt("Fml:Count", 0));
        Assert.Equal(1.5, configuration.GetDouble("Fml:Ratio", 0));
        Assert.Equal(250_000m, configuration.GetDecimal("Fml:Amount", 0m));
    }

    [Fact]
    public void Fallbacks_are_used_for_missing_and_unparseable_values()
    {
        var configuration = Configuration(("Fml:Count", "not-a-number"));

        Assert.True(configuration.GetBool("Fml:Missing", true));
        Assert.Equal(7, configuration.GetInt("Fml:Count", 7));
        Assert.Equal(2.0, configuration.GetDouble("Fml:Missing", 2.0));
        Assert.Equal(3m, configuration.GetDecimal("Fml:Missing", 3m));
    }

    [Theory]
    [InlineData("en-US")]
    [InlineData("de-DE")]
    [InlineData("fr-FR")]
    public void Numeric_values_are_parsed_independently_of_the_current_culture(string culture)
    {
        var previous = CultureInfo.CurrentCulture;
        CultureInfo.CurrentCulture = new CultureInfo(culture);
        try
        {
            var configuration = Configuration(
                ("Fml:Ratio", "1.5"),
                ("Fml:Amount", "250000.75"));

            Assert.Equal(1.5, configuration.GetDouble("Fml:Ratio", 0));
            Assert.Equal(250_000.75m, configuration.GetDecimal("Fml:Amount", 0m));
        }
        finally
        {
            CultureInfo.CurrentCulture = previous;
        }
    }
}
